### ISSUE TITLE
Tests fail

### ISSUE THREAD

**Issue by cyanate** [bus1/dbus-broker#19]
Which kernel version is needed at least to have SO_PEERGROUPS support?

I've compiled dbus-broker-git. Tests 28 and 34 failed. Here is my testlog:

[testlog.txt](https://github.com/bus1/dbus-broker/files/1246216/testlog.txt)

**Comment by teg** [comment]
Thanks for the report.

SO_PEERGROUPS was introduced in this commit (so most distro kernels will not have it yet): <https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/core/sock.c?id=28b5ba2aa0f55d80adb2624564ed2b170c19519e>.

**Comment by dvdhrm** [comment]
`SO_PEERGROUPS` is not mandatory, yet. You can ignore those warnings.

However, this part is interesting:

```
177 --- command ---                                                                  
178 ./dbus-broker-git/src/build/src/test-sasl                                        
179 --- stderr ---                                                                   
180 test-sasl: ../dbus-broker-git/src/dbus/test-sasl.c:157: test_client_run: Assertion `!r' failed.
```

It suggests that the client-sasl code fails (which also explains the other failure you see). Can you give some details how to reproduce this? The code there is rather trivial, so I really wonder what is going wrong there.

What machine do you run on? What linux distribution? Did you use one of the prebuilt packages or compile dbus-broker yourself? etc. etc.

Thanks!

**Comment by NikitaPuzyryov** [comment]
I had this issue when building with `-O3`. With the default `-O2` tests run successfully.

**Comment by dvdhrm** [comment]
Indeed. This fails with `-O3`. I will look into this.

**Comment by dvdhrm** [comment]
Oh god!

So this tracks down to:

```c
*arg = memchr(input, ' ', n_input);
```

Even if `n_input` is 0, you must make sure `input` is non-NULL. There is no way `memchr()` fails, but apparently gcc follows the C-standard strictly in this case, and then assumes `input` cannot be NULL. Hence, any following check for `if (input) ...` is silently optimized away, since how could `input` ever be NULL if you passed it to `memchr()`?

*sigh*

I can fix this one case, but I honestly don't expect our code-base to be safe against this insanity. I kinda prefer `Don't use -O3!`...

Comments?

**Comment by dvdhrm** [comment]
This is now fixed in -master via c5c8589c8c4e4dc2c3a23aa4edc840c0784b238a.

Thanks a lot for the report! I still recommend -O2 over -O3...
